bindings/c/test/mako/mako.cpp

```diff
@@ -581,70 +603,76 @@ int runWorkload(Database db,
 	/* main transaction loop */
 	while (1) {
 		Transaction tx = createNewTransaction(db, args, -1, args.active_tenants > 0 ? tenants : nullptr);
-		while ((thread_tps > 0) && (xacts >= current_tps)) {
+		if ((thread_tps > 0 /* iff throttling on */) && (xacts >= current_tps)) {
 			/* throttle on */
-			const auto time_now = steady_clock::now();
-			if (toDoubleSeconds(time_now - time_prev) >= 1.0) {
-				/* more than 1 second passed, no need to throttle */
-				xacts = 0;
-				time_prev = time_now;
-
-				/* update throttle rate */
-				current_tps = static_cast<int>(thread_tps * throttle_factor.load());
-			} else {
+			auto time_now = steady_clock::now();
+			while (toDoubleSeconds(time_now - time_prev) < 1.0) {
 				usleep(1000);
+				time_now = steady_clock::now();
 			}
+
+			/* more than 1 second passed*/
+			xacts = 0;
+			time_prev = time_now;
+
+			/* update throttle rate */
+			current_tps = static_cast<int>(thread_tps * throttle_factor.load());
 		}
-		/* enable transaction trace */
-		if (dotrace) {
-			const auto time_now = steady_clock::now();
-			if (toIntegerSeconds(time_now - time_last_trace) >= 1) {
-				time_last_trace = time_now;
-				traceid.clear();
-				fmt::format_to(std::back_inserter(traceid), "makotrace{:0>19d}", total_xacts);
-				logr.debug("txn tracing {}", traceid);
-				auto err = Error{};
-				err = tx.setOptionNothrow(FDB_TR_OPTION_DEBUG_TRANSACTION_IDENTIFIER, toBytesRef(traceid));
-				if (err) {
-					logr.error("TR_OPTION_DEBUG_TRANSACTION_IDENTIFIER: {}", err.what());
+
+		if (current_tps > 0 || thread_tps == 0 /* throttling off */) {
```

Comment:
It seemed cleaner before when there was "by this code line, throttling is either disabled or it is enabled and we have some TPS to work with", so that we didn't have to drag down TPS condition checks here. Could you explain why this was necessary?

Reply:
I think the main problem was that the ending condition was never checked in the previous throttling while loop. If  we have `current_tps = 0`, `signal.load() = SIGNAL_RED`,`throttle_factor.load() = 0`, and the process that updates the throttle factor has finished,  the current thread will stay in its throttling while loop forever. So we either have to do the ending condition check in the throttling while loop or do something like I did. Maybe it would be nicer to move the ending condition check to the beginning of the `while(1)` loop and work with `continue`s.